### DISCUSSION TITLE
Avoid duplicate style in catalogue json for  wfs service

### DISCUSSION THREAD

**Discussion by luca-moscatelli** [TerriaJS/terriajs#7376] [Q&A]
# Avoid redundant style duplication for WFS layers in TerriaJS configuration

Hi TerriaJS team,

Currently, when configuring **WFS layers** in the catalog, I have to define a separate style for each individual layer ID. This process involves a lot of repetitive code, as I need to duplicate the same style object with minor adjustments (e.g., the `id` field) for each layer.

For example:
```json
{
  "id": "layer1",
  "type": "wfs-group",
  "itemProperties": {
    "styles": [
      {
        "id": "style1",
        "title": "Style 1",
        "color": {
          "mapType": "enum",
          "enumColors": [
            {"value": "1", "color": "#0000FF"},
            {"value": "2", "color": "#00FF00"}
          ]
        },
        "hidden": false
      }
    ],
    "activeStyle": "style1"
  }
},
{
  "id": "layer2",
  "type": "wfs-group",
  "itemProperties": {
    "styles": [
      {
        "id": "style2",
        "title": "Style 2",
        "color": {
          "mapType": "enum",
          "enumColors": [
            {"value": "1", "color": "#0000FF"},
            {"value": "2", "color": "#00FF00"}
          ]
        },
        "hidden": false
      }
    ],
    "activeStyle": "style2"
  }
}
```

This approach becomes cumbersome when working with a large number of WFS layers that all share similar or identical styles. 

Is there any approach in the catalog that allows me to avoid all this duplicate code?

**Comment by zoran995** [comment]
Hi @luca-moscatelli, currently, there is no easy way to reuse configuration between catalog items. You might try a workaround using one root catalog group and defining `itemPropertiesByType` on it. It could work as you are mainly using `WFS`.
It would look something like this
```
"catalog": [
  {
    "type": "group",
    "name": "Custom WFS Group",
    "itemPropertiesByType": [
      {
        "type": "wfs",
        "itemProperties": {
          "styles": [
            {
              "id": "style1",
              "title": "Style 1",
              "color": { "colorColumn": "", "nullColor": "#ff0000" },
              "hidden": false
            },
            {
              "id": "style2",
              "title": "Style 2",
              "color": {
                "mapType": "constant",
                "colorColumn": "",
                "nullColor": "#0000ff"
              },
              "hidden": false
            }
          ]
        }
      }
    ],
    "members": [
      {
        "type": "wfs-group",
        "name": "test",
        "url": "https://services.ga.gov.au/gis/mineralogy/wfs",
        "itemProperties": {
          "activeStyle": "style1"
        }
      },
      {
        "type": "wfs-group",
        "name": "test2",
        "url": "https://services.ga.gov.au/gis/mineralogy/wfs",
        "itemProperties": {
          "activeStyle": "style2"
        }
      }
    ]
  }
],
```

You should be able to define most of the styling for later use and then apply it on a subgroup/layer basis. Yeah, it can still become very complex, but for now, it is the only idea I have for tackling this. I will let you know if I think of something easier/better.

Related issue https://github.com/TerriaJS/terriajs/issues/3384

**Reply by zoran995** [reply]
After more thinking and discussion internally, this doesn't look like something that should handled inside TerriaJS but more for a configuration app which allows reuse styling and generates catalog definition files on the fly. @luca-moscatelli let us know if we can help with anything else
